feat: Add session share link with more options

andre-code opened the pull request:
- Added button to create a share link for sessions.
- Added handler when the branch name or commit key is not valid or incomplete

### Successful case

![ok-session-share-link](https://user-images.githubusercontent.com/43388408/146211423-79e996c7-01eb-4717-8707-fcae84c666ea.gif)


### When the commit or brach name is invalid or incomplete

![Screenshot 2021-12-15 at 16 01 25](https://user-images.githubusercontent.com/43388408/146210587-3cc29d1a-7cba-4c40-a73e-b723d667d1fc.png)

Closes #1596

/deploy renku=ui1608-tests-sessions

lorenzo-cavazzi reviewed (changes requested):
Very nice feature!
I found a problem when playing around. If the url contains a wrong commit or branch (e.g. deleted branch/commit on the project, or just an error on copy/paste by the user), the UI seems to be stuck on the "Starting session" page.
It would be nice giving feedback to the user.

![Screenshot_20211214_140508](https://user-images.githubusercontent.com/43481553/146011397-4dc9f6aa-461d-417c-b167-7868f2c2acf8.png)

andre-code commented:
> Very nice feature! I found a problem when playing around. If the url contains a wrong commit or branch (e.g. deleted branch/commit on the project, or just an error on copy/paste by the user), the UI seems to be stuck on the "Starting session" page. It would be nice giving feedback to the user.
> 

Thanks for the suggestion, I added the feedback when the commit key or branch is not valid 😉

lorenzo-cavazzi reviewed (changes requested):
I fear this solution may have a few issues on projects with many branches/commits (check the inline content).
Instead of using the `force` parameter on `refreshCommits`, the fetch should either include the target branch the first time (not sure if possible, maybe it's triggered somewhere else) or wait until it's finished and then fetch again if a different target branch was defined in the autostart.

I think it may be necessary (or at least much easier) doing that check in the `triggerAutoStart` function instead of modifying the fetch/selectBranch and fetch/selectCommit functions .
The `triggerAutoStart` function already contains a 1000ms timeout to check if every requirement for the autostart is ready. You could use that to check if
* the branches/commits are not being fetched (otherwise wait for the next loop)
* check if an autostart branch has been defined and the currently selected branch is that one. Otherwise set it using the `selectBranch` function, or create the error if not available (mind the previous condition -- all the other fetches should not be ongoing)
* same check for the commit, using `selectCommit` -- after the previous `selectBranch` it would require a few loops since the commits fetching would be ongoing again

andre-code commented on client/src/notebooks/Notebooks.container.js:
```diff
@@ -367,29 +385,37 @@ class StartNotebookServer extends Component {
+  async refreshCommits(force = false) {
     if (this._isMounted) {
-      if (!this.projectModel.get("commits.fetching")) {
+      if (!this.projectModel.get("commits.fetching") || force) {
```
Good point 🙌, I find it effective to avoid fetching commits from the default branch when there is an autostart session so I added that validation in the View component.

In case there is no autostart session, the commits are retrieved from the default branch (current behavior).
In case of an invalid branch in the autostart session, it displays an error, stops the autostart, and retrieves the commits from the default branch so that all the features that depend on those commits continue working.

Also to improve the behavior of the starting session progress bar, I added the validation to don't fetch the notebooks before setting the branch and commit when is an autostart session.

lorenzo-cavazzi commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -2059,6 +2074,74 @@ class CheckNotebookIcon extends Component {
+    if (!data.namespace || !data.path)
+      return;
+    let urlSession = Url.get(Url.pages.project.session.autostart, data, true);
+    const parameters = JSON.parse(props.notebooks.lastParameters);
```
No need to convert back that stringified object from `lastParameters` to get branch and commit. That prop is meant to be used only by the `NotebooksCoordinator` to discard outdated data.

You can used `filters` here: `props.filters.branch.name` and `props.filters.commit.id`.

andre-code replied:
sure, updated

lorenzo-cavazzi reviewed (changes requested):
Great, your solution is better than the suggested one!

There are a couple of minor details to iron out before merging:
* When the autostart fails, the page keeps showing `Verifying available session` without retrieving them. [EXAMPLE](https://renku-ci-ui-1608.dev.renku.ch/projects/lorenzo.cavazzi.tech/project-renku-0-16-2/sessions/new?autostart=1&branch=abc)
![Screenshot_20211217_121841](https://user-images.githubusercontent.com/43481553/146537001-83a27373-866f-4687-bbfc-57deecbbaebc.png)
* When an autostart successfully completes, the page shows a message about "session not available" for a few seconds.
[EXAMPLE](https://renku-ci-ui-1608.dev.renku.ch/projects/lorenzo.cavazzi.tech/project-renku-0-16-2/sessions/new?autostart=1)
![Screenshot_20211217_111757](https://user-images.githubusercontent.com/43481553/146536721-4eaccd90-d7f6-4553-99b0-2ea3e768ae79.png)
It would be better to show the `<Loader>` -- I think that was the case before [EXAMPLE](https://dev.renku.ch/projects/lorenzo.cavazzi.tech/project-renku-0-16-2/sessions/new?autostart=1
)
![Screenshot_20211217_112935](https://user-images.githubusercontent.com/43481553/146536849-f886f9bb-4cc9-47f2-bffa-561095239779.png)

andre-code commented:
@lorenzo-cavazzi  Updated PR to start polling session if autostart fails (first case reported) and display the loader while getting session status (second case reported)

lorenzo-cavazzi commented on client/src/notebooks/Notebooks.present.js:
```diff
@@ -2056,6 +2071,78 @@ class CheckNotebookIcon extends Component {
+    urlSession = includeCommit ? `${urlSession}&commit=${data.commit}` : urlSession;
+    urlSession = includeBranch ? `${urlSession}&branch=${data.branch}` : urlSession;
+    setUrl(urlSession);
+  }, [ includeCommit, includeBranch, props.notebooks.lastParameters, data ]);
```
```suggestion
  }, [ includeCommit, includeBranch, data ]);
```

lorenzo-cavazzi reviewed (approved):
It looks good to me! :tada: 
Thanks for polishing all the details.

P.S. I'm re-running the acceptance tests, the previous errors were confusing since I could work with sessions and manually replicate the steps from the tests :shrug:
